reactor-core-micrometer/src/main/java/reactor/core/observability/micrometer/MicrometerObservationListener.java

```diff
@@ -0,0 +1,211 @@
+		subscribeToTerminalObservation = Observation.createNotStarted(
+			configuration.sequenceName + OBSERVATION_FLOW,
+			configuration.registry
+		).lowCardinalityKeyValues(configuration.commonKeyValues);
+	}
+
+	@Override
+	public void doOnCancel() {
+		//Due to dealing with scopes, we have to stop the current scope.
+		//Note this could skew the onNext counter off by one if the Observation is a facade over metrics.
+		if (this.onNextScope != null) {
+			this.onNextScope.close();
+		}
+
+		Observation observation = subscribeToTerminalObservation
+			.lowCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_STATUS, MicrometerMeterListener.TAG_STATUS_CANCELLED)
+			.highCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_EXCEPTION, "");
+
+		observation.stop();
+	}
+
+	@Override
+	public void doOnComplete() {
+		//Due to dealing with scopes, we have to stop the current scope.
+		//Note this could skew the onNext counter off by one if the Observation is a facade over metrics.
+		if (this.onNextScope != null) {
+			this.onNextScope.close();
+		}
+
+		// We differentiate between empty completion and value completion via tags.
+		String status = null;
+		if (!valued) {
+			status = MicrometerMeterListener.TAG_STATUS_COMPLETED_EMPTY;
+		}
+		else if (!configuration.isMono) {
+			status = MicrometerMeterListener.TAG_STATUS_COMPLETED;
+		}
+
+		// if status == null, recording with OnComplete tag is done directly in onNext for the Mono(valued) case
+		if (status != null) {
+			Observation completeObservation = subscribeToTerminalObservation
+				.lowCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_STATUS, status)
+				.highCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_EXCEPTION, "");
+
+			completeObservation.stop();
+		}
+	}
+
+	@Override
+	public void doOnError(Throwable e) {
+		//Due to dealing with scopes, we have to stop the current scope.
+		//Note this could skew the onNext counter off by one if the Observation is a facade over metrics.
+		if (this.onNextScope != null) {
+			this.onNextScope.close();
+		}
+
+		Observation errorObservation = subscribeToTerminalObservation
+			.lowCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_STATUS, MicrometerMeterListener.TAG_STATUS_ERROR)
+			.highCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_EXCEPTION, e.getClass().getName());
+
+		errorObservation.stop();
+	}
+
+	@Override
+	public void doOnNext(T t) {
+		valued = true;
+		if (signalIntervalObservation == null || onNextScope == null) { //NB: interval observation is only null if isMono
+			//record valued completion directly
+			Observation completeObservation = subscribeToTerminalObservation
+				.lowCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_STATUS, MicrometerMeterListener.TAG_STATUS_COMPLETED)
+				.highCardinalityKeyValue(MicrometerMeterListener.TAG_KEY_EXCEPTION, "");
+
+			completeObservation.stop();
+			return;
+		}
+		//record the delay since previous onNext/onSubscribe. This also records the count.
+		this.onNextScope.close();
```

Comment:
I don't entirely understand this. When you close the scope we automatically make the previous one the current, so you don't need to manually open a new scope.